Update CONTRIBUTING.md (#1379)
jamulussoftware/jamulus: CONTRIBUTING.md

- Please only start coding after we have agreed to a specification.

## Jamulus project/source code general principles

### Stability
This has been, and must continue to be, the most important requirement. It is particularly important for use cases such as the Worldjam where instability during a live performance would be unacceptable. The following two principles are designed to support this principle.

### [Keep it Simple and Stupid](https://en.wikipedia.org/wiki/KISS_principle) and [Do One Thing and Do It Well](https://en.wikipedia.org/wiki/Unix_philosophy#Do_One_Thing_and_Do_It_Well)
If a feature or function can be achieved in another way by another system or method, then that is preferable to building that feature into Jamulus. 

### Focus on and improve the strengths of Jamulus
There are other solutions for real time online jamming available. But Jamulus stands alone with its client/server system, minimalistic GUI and simple functionality. 

### Source code consistency

> Respect the existing code style: Tab size=4, insert spaces.

Reviewer: Are there any more detailed coding style guidelines? Or is it just a case of copying what's there? For example:
1. Space before and after `(` and `)`, _except_ no space between `)` and `;`, and no space before an empty `()`.
2. All bodies of `if`, `else`, `while`, `for`, etc., to be enclosed in braces `{` and `}`, usually on separate lines.
3. Other conventions I can't think of offhand.

Author: This was the only stuff I could find (originally on SF, then GH Discussion). Some mention of creating more, but I don't think it materialised?

Make sure (if possible) that your code compiles on Windows/Mac/Linux.
Do not use diff/patch to send your code changes but create a Github fork of the Jamulus code and create a Pull Request when you are done.

### Documentation/Acknowledgements

> Each new feature or bug fix must be documented in the ChangeLog. If a new contributor/translator adds code to the Jamulus project, he/she should be added to the contributor/translator list in the About dialog of the Jamulus software.

Reviewer: Mention that this list is in `src/util.cpp` in the constructor function `CAboutDlg::CAboutDlg()`.

Reviewer: There is the tricky issue of multiple people adding themselves to `src/util.cpp` in different PRs, resulting in merge conflicts. We've already recently found this with `ChangeLog` too. Volker used to maintain the change log and the contributors lists himself, probably for this reason.

### Merging Pull Requests

> The git master branch is protected so that at least two reviews must be done before the pull request can be merged. If a pull request received at least two positive reviews, any of the main developers can initiate the merge.

Reviewer: "...  reviews by at least two of the main developers must be done ..."

# Want to contribute some code? 

## We’d really appreciate your support!

- If the [Github issue](https://github.com/jamulussoftware/jamulus/issues) for your feature/bug fix already exists, write a message in that issue saying you want to work on it. 
- Otherwise, please [post on the GitHub Discussions](https://github.com/jamulussoftware/jamulus/discussions) and say that you are planning to do some coding and explain why. Then we can discuss the specification. 
- Please only start coding after we have agreed to a specification.

## Jamulus project/source code general principles

### Stability
This has been, and must continue to be, the most important requirement. It is particularly important for use cases such as the Worldjam where instability during a live performance would be unacceptable. The following two principles are designed to support this principle.

### [Keep it Simple and Stupid](https://en.wikipedia.org/wiki/KISS_principle) and [Do One Thing and Do It Well](https://en.wikipedia.org/wiki/Unix_philosophy#Do_One_Thing_and_Do_It_Well)

> If a feature or function can be achieved in another way by another system or method, then that is preferable to building that feature into Jamulus. 

Reviewer 2: For me, this sounds a bit strange. Doesn't it mean the contrary?

Author: Probably too easy to miss the "that" referring to "in another way". Maybe clearer as:

"If a feature or function can be achieved by other means, then those things should not be built into Jamulus."

Reviewer 3: Volker Fischer only wrote "keep it simple" in his paper. ;-)

Reviewer 2: Yep. This sounds clearer

Author: @jujudusud yes, and later expanded that https://github.com/jamulussoftware/jamulus/discussions/915

### Focus on and improve the strengths of Jamulus

> There are other solutions for real time online jamming available. But Jamulus stands alone with its client/server system, minimalistic GUI and simple functionality. 

Reviewer 4: This is the "Why Jamulus rather than ..?" quiz :).

I've not heard much comparative discussion of Jamulus against other realtime jamming solutions in terms of the user experience, either - at least, not by dispassionate parties without specific allegiances...

This is written to the developer, though.  But even there, "simple" it isn't...  I think this was trying to say that the aim was "Run it, click connect, jam" - simple.  Just not saying "Run in.  Work out what Soundcard Device I'm using.  Work out why it doesn't like it.  Work out whether I need to worry about all those other settings.  Click connect.  Wait.  Why did only a handful of servers and not the one I was told to pick show up?"

;)

> There are other solutions for real time online jamming available. But Jamulus stands alone with its client/server system, minimalistic GUI and simple functionality. 

Author: It feels like a re-statement of the "KISS" principle in the sense that it emphasises minimalism and simplicity. So either we remove it as overlapping too much, or think of another principle that makes Jamulus special. But as you say, nothing immediately comes to mind.

> There are other solutions for real time online jamming available. But Jamulus stands alone with its client/server system, minimalistic GUI and simple functionality. 

Reviewer 5: I'd suggest removing it for now. It will be hard enough to get contributors to read the whole document. It'll be simpler if it's shorter and avoids redundancy.

> There are other solutions for real time online jamming available. But Jamulus stands alone with its client/server system, minimalistic GUI and simple functionality. 

Author: OK. And if we think of any other principle to live by, we can always add it.

### Source code consistency
- Respect the existing code style: Tab size=4, insert spaces.
- Space before and after `(` and `)`, except no space between `)` and `;`, and no space before an empty `()`.

> - All bodies of `if`, `else`, `while`, `for`, etc., to be enclosed in braces `{` and `}`, usually on separate lines.

Reviewer 5: ```suggestion
- All bodies of `if`, `else`, `while`, `for`, etc., to be enclosed in braces `{` and `}`, on separate lines.
```

Reviewer 4: Not always.  In a header, it's common to see a one line method definition.  Indeed, it's the preferred style for methods in headers -- but really they shouldn't have `if` statements in (and I always feel so guilty adding them and not moving the body to the `.cpp` file...).

Reviewer 4: One exception: methods in headers.
```
Where a method has only a single statement, e.g. a getter or setter with no logic, that can be placed in the header declaration on the same line as the method name (subject to wrapping considerations).  There is no requirement to split with the braces in this case.
```

> Make sure (if possible) that your code compiles on Windows/Mac/Linux.

Reviewer 4: Check your Github fork.  It should have built your branch for you when you pushed, for all the supported platforms (though not all supported platform versions).  Make sure there are no errors.

Reviewer 5: > Check your Github fork. It should have built your branch for you when you pushed, for all the supported platforms (though not all supported platform versions).

I think this is only true for branches starting with `autobuild/` (or `autobuild_`) or specific tags, isn't it?

Reviewer 4: Didn't think so.

See https://github.com/jamulussoftware/jamulus/pull/1370/checks for example.

> Do not use diff/patch to send your code changes but create a Github fork of the Jamulus code and create a Pull Request when you are done.

Reviewer 4: If there are conflicts with jamulussoftware/jamulus showing when you go to raise the pull request, resolve those locally:
```
git checkout master ;# checkout your local master
git pull upstream master --tags ;# refresh from jamulussoftware/jamulus (i.e. use your upstream remote name)
git push --tags --force ;# update your fork's master
git checkout - ;# back to your working branch
git rebase master ;# replay your changes onto the latest code
git push --force ;# update your working branch on your fork
```

> 

Reviewer 5: I think we should document our compatibility guarantees somewhere and I think this might be the best place?

I don't want to hijack this PR, so if you want to skip this for now, it is ok, then we should have a follow-up PR.

```suggestion
### Supported platforms
We strive to support the following platforms and versions:
- Windows 10 <!-- Is this the only version? Do we have to specify a specific sub-version? -->
- Mac <!-- Which versions? -->
- Ubuntu Linux 18.04 or later, Debian 10 or later, most Linux flavors with recent enough Qt versions
<!-- Do we support BSD? I think a recent discussion was about FreeBSD? -->
<!-- Should we already list Android? If so, what platforms/versions? -->

The minimum Qt version to support is 5.X <-- What is it? -->

Please try hard to avoid breaking any of them by avoiding to introduce platform-specific code and checking if any newly introduced Qt calls are available in our minimum Qt version.

### Dependencies
If your code requires new dependencies, be extra sure that those are available on all supported platforms and that the introduction of those has been discussed and accepted.
```

Reviewer 4: Windows 7 was what I thought we aimed for as a minimum version.  Later should "just work" with ASIO and Qt:
https://doc.qt.io/qt-5/windows.html

Basically, I think cross-platform, it's "Whatever Qt5 and our supported sound platform is support".

Reviewer: > Windows 7 was what I thought we aimed for as a minimum version. Later should "just work" with ASIO and Qt:
> https://doc.qt.io/qt-5/windows.html

Yes, I agree. I still use Windows 7 on my main PC, which is nearly 9 years old. I think we would need to have a good reason to stop working on W7.

> I think some (Mac?) platforms needed Qt5.9 for some reason...

Apparently Qt 5.9 supports back to macOS 10.10 (Yosemite), 5.10/5.11 support back to macOS 10.11 (El Capitan), 5.12/5.13 support back to macOS 10.12 (Sierra) and 5.14/5.15 support back to macOS 10.13 (High Sierra). We have no way of knowing how many users need support for those old versions. I use 10.14 (Mojave) on mine.

Reviewer 5: The reason for listing explicit (minimum) OS versions was to find out what the oldest supported Qt version is so that we can list it. If we just specify a minimum Qt version, it may be hard to raise that at some point.

> I think we need to start working on a Qt exit plan...

Should we have a Github discussion about that? I've only recently learned about these changes in Qt.

### Documentation/Acknowledgements

> Each new feature or bug fix must be documented in the ChangeLog. If a new contributor/translator adds code to the Jamulus project, he/she should be added to the contributor/translator list in the About dialog of the Jamulus software (see in `src/util.cpp` in the constructor function `CAboutDlg::CAboutDlg()`).

Reviewer 5: For 3.7.0 we asked contributors to create the Changelog entry. This has the downside of creating conflicts, which is annoying for other contributors. I suggest to change this to have people optionally suggest a changelog entry line in the PR description, but create the actual Changelog entries as part of the release process.

I think we can stick to people adding themselves to the contributors list because this is a one-time action per-person.

I also suggest to make it clearer whose job it is to do these things.

Suggestion:
```suggestion
Each new feature or bug fix must be documented in the ChangeLog. Please provide a single-sentence suggestion for that as part of your pull request description. Do not modify the ChangeLog file as part of your PR as it will lead to conflicts.

If you are a first-time contributor/translator, please add yourself to the contributor/translator list in the About dialog of the Jamulus software (see in `src/util.cpp` in the constructor function `CAboutDlg::CAboutDlg()`).
```

## Jamulus project/source code general principles

### Stability
This has been, and must continue to be, the most important requirement. It is particularly important for use cases such as the Worldjam where instability during a live performance would be unacceptable. The following two principles are designed to support this principle.

### [Keep it Simple and Stupid](https://en.wikipedia.org/wiki/KISS_principle) and [Do One Thing and Do It Well](https://en.wikipedia.org/wiki/Unix_philosophy#Do_One_Thing_and_Do_It_Well)
If a feature or function can be achieved in another way by another system or method, then that is preferable to building that feature into Jamulus. 

### Source code consistency
- Respect the existing code style: Tab size=4, insert spaces.
- Space before and after `(` and `)`, except no space between `)` and `;`, and no space before an empty `()`.
- All bodies of `if`, `else`, `while`, `for`, etc., to be enclosed in braces `{` and `}`, on separate lines.

Do not use diff/patch to send your code changes but create a Github fork of the Jamulus code and create a Pull Request when you are done.

> Please check your Github fork. It should build your branch for all the supported platforms (though not all supported platform versions) when you push. Make sure there are no errors. If there are, resolve them before creating a pull request.

Reviewer 5: Are we sure that this is possible?
I suspect the autobuild only runs for release tags, on master and on PRs. This would mean that no autobuild runs on your own fork's feature branches.

Author: Volker wrote that, but I guess it's possible he never tried it?

Reviewer 5: I just checked: As I suspected, there is no autobuild for this case.

Here's a feature branch of mine in my fork: https://github.com/hoffie/jamulus/tree/recorder-cleanup
The `Actions` tab does not list any autobuild jobs within the last hours.

The PR (#1463) has the proper CI jobs.

```suggestion
Please run a local build test. Make sure there are no errors. After opening a pull request, keep an eye on the CI checks for quality or compile issues and fix them as required.
```

As an alternative, we could change the autobuild logic (@ann0see @nefarius2001), if that's wanted.

Reviewer 4: I think every push should trigger a build.  (It can quite happily kill older builds on the branch, too, if that's feasible.)

If there are conflicts with jamulussoftware/jamulus showing when you go to raise the pull request, resolve those locally:
~~~
git checkout master ;# checkout your local master
git pull upstream master --tags ;# refresh from jamulussoftware/jamulus (i.e. use your upstream remote name)
git push --tags --force ;# update your fork's master
git checkout - ;# back to your working branch
git rebase master ;# replay your changes onto the latest code
git push --force ;# update your working branch on your fork
~~~

### Supported platforms
We support the following platforms and versions:

> - Windows 10 <!-- Is this the only version? Do we have to specify a specific sub-version? -->

Reviewer 5: ```suggestion
- Windows 7, 10 <!-- Is this the only version? Do we have to specify a specific sub-version? -->
```

as per previous comments.

Reviewer 2: Windows 8.1 works too. I'd say 7 or higher.

> - macOS <!-- Which versions? -->

Reviewer 5: Someone should fill that in :)

- Ubuntu Linux 18.04 or later, Debian 10 or later, most Linux flavors with recent enough Qt versions
<!-- Do we support BSD? I think a recent discussion was about FreeBSD? -->
<!-- Should we already list Android? If so, what platforms/versions? -->

> The minimum Qt version to support is 5.X <-- What is it? -->

Reviewer 5: ```suggestion
The minimum Qt version to support is 5.X <!-- What is it? -->
```

Still don't know what to fill in, but this should be documented, IMO. :(

Reviewer 4: We support 5.12.2 on all platforms except MacOS, where we need to support 5.9 -- so that means Qt 5.9 as a latest version for build.

It should be noted that _code style_ in a file may be Qt 4.x.  Normally, "stick to existing style" applies.  However, if making large-scale changes, then updating to Qt 5.9 is encouraged.

- Space before and after `(` and `)`, except no space between `)` and `;`, and no space before an empty `()`.
- All bodies of `if`, `else`, `while`, `for`, etc., to be enclosed in braces `{` and `}`, on separate lines.

Do not use diff/patch to send your code changes but create a Github fork of the Jamulus code and create a Pull Request when you are done.

Please run a local build test. Make sure there are no errors. After opening a pull request, keep an eye on the CI checks for quality or compile issues and fix them as required.

If there are conflicts with jamulussoftware/jamulus showing when you go to raise the pull request, resolve those locally:
~~~
git checkout master ;# checkout your local master
git pull upstream master --tags ;# refresh from jamulussoftware/jamulus (i.e. use your upstream remote name)
git push --tags --force ;# update your fork's master
git checkout - ;# back to your working branch
git rebase master ;# replay your changes onto the latest code
git push --force ;# update your working branch on your fork
~~~

### Supported platforms
We support the following platforms and versions:
- Windows 7 or higher

> - macOS <!-- Which versions? -->

Reviewer 2: According to https://jamulus.io/wiki/Getting-Started it's OS X 10.11?

Author: Yes, not sure where that came from. I think it was a report from a Mac user who had 10.10?

Reviewer 2: Or maybe it's Qt 5.9? No: https://doc.qt.io/archives/qt-5.9/supported-platforms.html

Author: Maybe - so shall we stick with 10.11? Feels like we almost need a variable for the supported version of the OS :-)

Reviewer 2: Yes. Probably add 10.11

- Ubuntu Linux 18.04 or later, Debian 10 or later, most Linux flavors with recent enough Qt versions
<!-- Do we support BSD? I think a recent discussion was about FreeBSD? -->

> <!-- Should we already list Android? If so, what platforms/versions? -->

Reviewer 2: We could list Android (and probably iOS soon). But I have no idea which version.

Author: I don't think we can say we "support" those those in the general sense of doing releases yet though, can we?

Reviewer 2: Hmm. Android maybe but iOS no.

Reviewer 5: I think there are two perspectives:
- From a developer point of view (which should matter here), we should avoid breaking the Android/iOS builds if possible. So in a way, we should "support" it.
- From a user's perspective, we should declare Android (and especially iOS) as experimental/unsupported until it is ready for prime time.

Please try to avoid breaking any of them by introducing platform-specific code. Check to see if any newly introduced Qt calls are available in the minimum supported Qt version, which is currently 5.9. Note that code _style_ in a file may be Qt 4.x, and while you should normally stick to existing style if making large-scale changes, then updating to Qt 5.9 style is encouraged.

> 

Reviewer 5: Based on @softins's https://github.com/jamulussoftware/jamulus/pull/963#issuecomment-817202182:
- Requiring Qt 5.9 seems correct.
- We should also document C++11.

```suggestion
Keep all code compatible to C++11.
```

> Keep all code compatible to C++11.

Reviewer 4: Blank line before and after this?

### Dependencies
If your code requires new dependencies, be sure that those are available on all supported platforms and that the introduction of those dependencies has been discussed and accepted.

### Ownership
The submitter of an Issue or a PR is responsible for its care and feeding, answering all questions directed at them, and making agreed amends if needed. 

So as to reduce effort for others in understanding a PR or an Issue, authors are strongly encouraged to update their initial posts/PR descriptions or title to reflect the current state of play, amends, enhancements, outstanding issues etc. Admins reserve the right to do this as they see fit.

### Documentation/Acknowledgements
Each new feature or bug fix must be documented in the ChangeLog. Please provide a single-sentence suggestion for that as part of your pull request description. Do not modify the ChangeLog file as part of your PR as it will lead to conflicts.

> If you are a first-time contributor/translator, please add yourself to the contributor/translator list in the About dialog of the Jamulus software (see in `src/util.cpp` in the constructor function `CAboutDlg::CAboutDlg()`).

Reviewer 2: Just thinking out loud: Adding a link would break if we make changes to this file. Nevertheless, new contributors would need to scroll a lot here.

Reviewer 5: Also thought about it. We could add a permalink to a specific revision to point to the right place. This will continue working, but might be confusing on the other hand. :(

Author: Now that I've done this https://github.com/jamulussoftware/jamuluswebsite/issues/314 I think I'll merge this too for consistency, and raise an issue about this point somewhere.

### Dependencies
If your code requires new dependencies, be sure that those are available on all supported platforms and that the introduction of those dependencies has been discussed and accepted.

### Ownership
The submitter of an Issue or a PR is responsible for its care and feeding, answering all questions directed at them, and making agreed amends if needed. 

> So as to reduce effort for others in understanding a PR or an Issue, authors are strongly encouraged to update their initial posts/PR descriptions or title to reflect the current state of play, amends, enhancements, outstanding issues etc. Admins reserve the right to do this for them if need be.

Reviewer 4: If change "if need be" to "as they see fit" :).
